broadinstitute/single_cell_portal_core
# Improve Image Pipeline scalability: use prefetched expression (SCP-4563)

This enables Image Pipeline to retrieve expression data from local files, so it processes more robustly and faster.

### Impact
[Previously](#1577), expression data for gene expression scatter plot images was fetched via SCP REST API in Image Pipeline.  This was convenient, but scaled poorly.  It entailed immense load on the SCP application and database servers -- enough so that we'd need to do something like transiently shard MongoDB, spin up infrastructure copies via Terraform, and generally do a bunch of complicated stuff in order to not take down production.  In terms of performance, that complicated approach would at best still be rather slow for studies with extremely many cells.

Now, expression data can be read from local, prefetched files.  With the forthcoming scatter data pipeline, we'll be able to generate expression data arrays for all genes and clusters in a given study _without_ hitting SCP API and all its attendant load and speed issues.  Combined with that work, this enhancement will essentially eliminate "backend" and "transfer" time.  

That [saves > 10 seconds per gene](https://mixpanel.com/project/2120588/view/19059/app/insights/#report/31338756/expression-scatter-plot-time-0-5-m-1-m-points), which with 25K genes <ins>_saves almost 3 days per Image Pipeline run_</ins> on large studies.

### Architecture
Below is a diagram showing where this fits in the upcoming architecture.  This derisks some of the "Image Pipeline" part at right -- specifically, that it can read standalone scatter data files (much of the green arrow).  Forthcoming work in Ingest Pipeline will prove we can write such scatter data.  

[<img src="https://user-images.githubusercontent.com/1334561/187295866-2abcba4d-bfbd-4312-9ed5-99492e2df5fc.png" width="500" />](https://docs.google.com/document/d/1-mhtoWrg3RHoDjWrHGv2h-KtlnnYGPg5-lxlgRyWaK0/edit#heading=h.cg3befaxtgyj)

Future work will mature how Image Pipeline fetches scatter data from the bucket, handles file names, etc.

To test:
* Get the accession for your local "Human milk -- differential expression" study
* In Terminal, `cd image-pipeline`
* Substituting your accession, run the following command for 2 minutes, then terminate it:
`rm -rf output/; export NODE_TLS_REJECT_UNAUTHORIZED=0; time node expression-scatter-plots.js --accession="<your accession>"; tput bel; # localhost, human milk DE study`
  * If you immediately see `TypeError: fetch failed`, edit `expression-scatter-plots.js` to use `https://127.0.0.1:3000` instead of `https://localhost:3000`, then re-run above
* Confirm the following shows JSON files: `ls -latr output/<your_accession>/json/`
* Substituting your accession, run the following command for 2 minutes (or a few seconds after you see `Completed loading Explore tab` in console)
`export NODE_TLS_REJECT_UNAUTHORIZED=0; time node expression-scatter-plots.js --accession="<your accession>"; tput bel; # localhost, human milk DE study`
* While that is running, terminate your local `rails s` command and confirm the Image Pipeline script still runs
  * This can be a bit tricky, because your OS focus will be yanked back to Chromium.  Go to the terminal tab you're running `rails s` on then kill it via e.g.  `Ctrl + c`.

This satisfies SCP-4563.
